vlm/src/main/scala/LayoutTileSource.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2018 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.contrib.vlm
+
+import geotrellis.raster._
+import geotrellis.spark.tiling._
+import geotrellis.spark.{SpatialKey}
+
+/** Reads tiles by key from a [[RasterSource]] as keyed by a [[LayoutDefinition]]
+  * @note It is required that the [[RasterSource]] is pixel aligned with the [[LayoutDefinition]]
+  *
+  * @param source raster source that can be queried by bounding box
+  * @param layout definition of a tile grid over the pixel grid
+  */
+class LayoutTileSource(val source: RasterSource, val layout: LayoutDefinition) {
+  // TODO: require that source and layout are grid alligned, consider floating point precision
+  def sourceColOffset: Long = ((source.extent.xmin - layout.extent.xmin) / layout.cellwidth).toLong
+  def sourceRowOffset: Long = ((layout.extent.ymax - source.extent.ymax) / layout.cellheight).toLong
+
+  /** Read tile according to key.
+    * If tile area intersects source partially the non-intersecting pixels will be filled with NODATA.
+    * If tile area does not intersect source None will be returned.
+    */
+  def read(key: SpatialKey): Option[MultibandTile] = {
+    val col = key.col.toLong
+    val row = key.row.toLong
+    val sourcePixelBounds = GridBounds(
```

Comment:
Might be good to put this logic in its own private method.

Reply:
Hm, i'd say `readAll` can be the following: 

```scala
def readAll(keys: Iterator[SpatialKey]): Iterator[(SpatialKey, MultibandTile)] =
    keys.flatMap(k => read(k).map(v => k -> v))
```

Reply:
The reason it doesn't is because our GeoTiffReader does a lot better when it reads multiple GridBounds at the same time, which is the method that would be invoked here. I don't know how long we can maintain the distinction between "read one" and "read multiple" in the API though.

Reply:
`GridBounds` already has `offset` method that does this, just trying to save an extra class allocation here I guess?